modyn/selector/internal/selector_strategies/presampling_strategies/label_balanced_presampling_strategy.py

```diff
@@ -0,0 +1,141 @@
+from typing import Optional
+
+from modyn.metadata_database.metadata_database_connection import MetadataDatabaseConnection
+from modyn.metadata_database.models import SelectorStateMetadata
+from modyn.selector.internal.selector_strategies.presampling_strategies import AbstractPresamplingStrategy
+from sqlalchemy import Select, asc, func, select
+from sqlalchemy.sql.expression import func as sql_func
+
+
+def get_fair_share(capacity: int, requests: list[int]) -> int:
+    """
+    Simple Fair Sharing algorithm.
+    Here it's used to determine how many samples gets each class (min(fair_share, num_samples)
+    """
+    fair_share = int(capacity / len(requests))
+    below_fair_share = [requests.index(el) for el in requests if el <= fair_share]
+
+    if len(below_fair_share) != 0 and len(requests) != len(below_fair_share):
+        new_capacity = capacity - sum(requests[i] for i in below_fair_share)
+        new_requests = [requests[i] for i in range(len(requests)) if i not in below_fair_share]
+        return get_fair_share(new_capacity, new_requests)
+    return fair_share
+
+
+def get_fair_share_predicted_total(fair_share: int, requests: list[int]) -> int:
+    return sum(min(fair_share, req) for req in requests)
+
+
+class LabelBalancedPresamplingStrategy(AbstractPresamplingStrategy):
+    def __init__(self, presampling_config: dict, modyn_config: dict, pipeline_id: int, maximum_keys_in_memory: int):
+        super().__init__(presampling_config, modyn_config, pipeline_id, maximum_keys_in_memory)
+
+        self.force_required_target_size = presampling_config.get("force_required_target_size", False)
+        self.force_label_balancing = presampling_config.get("force_label_balancing", False)
+
+    def get_presampling_query(
+        self,
+        next_trigger_id: int,
+        tail_triggers: Optional[int],
+        limit: Optional[int],
+        trigger_dataset_size: Optional[int],
+    ) -> Select:
+        samples_count = self._get_samples_count_per_label(next_trigger_id, tail_triggers)
+        target_size = self.get_target_size(sum(samples_count), limit)
+        fair_share = get_fair_share(target_size, samples_count)
+
+        # randomly permute samples class by class. Then use row_number (in the next query) to get N random samples for
+        # each class
+        subquery = select(
+            SelectorStateMetadata,
+            sql_func.row_number()
+            .over(partition_by=SelectorStateMetadata.label, order_by=func.random())  # pylint: disable=not-callable
```

Comment:
Here the magic happens

Reply:
actually, we should use the number of classes for this pipeline. PR #269 (which should be merged today) will introduce a `num_classes` parameter. This way, if you actually have 3 classes, but only have received data of 2, we will not assume there are only 2 classes

Reply:
Why do you think it's needed? I mean, if a class has 0 samples (so it's ignored by the query above) you can either get 0 samples for each class (useless) or get fair_share samples (so the empty class gets zero data points)

Reply:
ok, true, agreed.